File: apps/backend/app/schemas/practice.py
from datetime import datetime

from pydantic import BaseModel, Field


class GenerateProblemRequest(BaseModel):
    problem_type: str = Field(pattern="^(quiz|coding)$")
    difficulty: str = Field(pattern="^(easy|medium|hard)$")
    tags: list[str] = Field(default_factory=list, max_length=5)
    source_lesson_id: str | None = None


class GenerateProblemResponse(BaseModel):
    task_id: str
    status: str = "pending"
    message: str = "問題を生成中です"


class TaskStatusResponse(BaseModel):
    task_id: str
    status: str
    problem: dict | None = None
    error_message: str | None = None


class SubmitPracticeRequest(BaseModel):
    problem_id: str
    submitted_code: str | None = None
    answer: dict | None = None
    time_spent_seconds: int | None = None


class SubmitPracticeResponse(BaseModel):
    is_correct: bool
    score: int
    ai_review: str | None = None
    ai_review_task_id: str | None = None

File: packages/shared-types/src/practice.ts
import type { AsyncTaskStatus } from "./common";

export type ProblemType = "quiz" | "coding";
export type Difficulty = "easy" | "medium" | "hard";

export type GenerateProblemRequest = {
  problem_type: ProblemType;
  difficulty: Difficulty;
  tags?: string[];
  source_lesson_id?: string | null;
};

export type GenerateProblemResponse = {
  task_id: string;
  status: AsyncTaskStatus;
  message: string;
};

export type TaskStatusResponse = {
  task_id: string;
  status: AsyncTaskStatus;
  problem: Record<string, unknown> | null;
  error_message: string | null;
};

export type SubmitPracticeRequest = {
  problem_id: string;
  submitted_code?: string | null;
  answer?: Record<string, unknown> | null;
  time_spent_seconds?: number | null;
};

export type SubmitPracticeResponse = {
  is_correct: boolean;
  score: number;
  ai_review: string | null;
  ai_review_task_id: string | null;
};
